### PR TITLE
Check type declarations with JSDoc comments

### LINKED ISSUE #8: Export TS types
Would be helpful 😀

### PR DESCRIPTION
**Pull request by colinrotherham**
Great work on Prettier v3 🙌

Hope this PR is welcome, but I've added JSDoc + **tsconfig.json** to export type declarations

I followed the link in the [**Prettier 3.0: Hello, ECMAScript Modules!** blog post](https://prettier.io/blog/2023/07/05/3.0.0.html) after realising that the Prettier sync APIs would still be needed to avoid [Nunjucks template changes for asynchronous rendering](https://mozilla.github.io/nunjucks/api.html#asynchronous-support) 😣

>If you still need sync APIs, you can try [@prettier/sync](https://github.com/prettier/prettier-synchronized)

I'd be happy to make any tweaks, and wasn't sure if committing the types would be preferred

Closes: https://github.com/prettier/prettier-synchronized/issues/8

### CODE REVIEW

**Comment by fisker** [comment]
Thanks for working on this. Can it be done by adding a `.d.ts` file?

**Comment by colinrotherham** [comment]
> Thanks for working on this. Can it be done by adding a `.d.ts` file?

Yeah course, I've pushed up a new **.gitignore** to show them:

1. Added `index.d.cts` type declarations for `.cjs` extension files
2. Added `index.d.cts.map` type declaration map to link back to source

>[**Declaration Map**](https://www.typescriptlang.org/tsconfig#declarationMap)
>Generates a source map for `.d.ts` files which map back to the original `.ts` source file. This will allow editors such as VS Code to go to the original `.ts` file when using features like _Go to Definition_.

I've used `import('prettier')` in the JSDoc to make sure the [existing type declarations](https://github.com/prettier/prettier/issues/14033) are reused when built

Imports types from Prettier

File: index.d.cts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  "formatWithCursor",
+  "format",
+  "check",
+  "resolveConfig",
+  "resolveConfigFile",
+  "clearConfigCache",
+  "getFileInfo",
+  "getSupportInfo",
+];
+declare const PRETTIER_STATIC_PROPERTIES: readonly ["version", "util", "doc"];
```
Comment (fisker) [inline index.d.cts]: I'm not familiar with types, but why are these needed?

File: index.d.cts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+declare const _exports: PrettierSync;
+export = _exports;
+export type Prettier = typeof import("prettier");
```
Comment (fisker) [inline index.d.cts]: Is it possible to do something like

```js
import type Prettier from "prettier";
```

?

File: package.json [hunk 3]
```diff
@@ -16,18 +16,29 @@
   },
   "sideEffects": false,
   "type": "module",
-  "exports": "./index.cjs",
+  "main": "./index.cjs",
+  "exports": {
+    ".": {
+      "types": "./index.d.cts",
+      "require": "./index.cjs",
+      "default": "./index.cjs"
```
Comment (fisker) [inline package.json]:
```suggestion
      "types": "./index.d.cts",
      "default": "./index.cjs"
```

Comment (colinrotherham) [reply package.json]: I'm not 100% certain about this

Not sure all runtimes and bundlers fall back to `"default"` when `"require"` (or `"import"`) is missing

Node.js resolution works of course:

```shell
node --eval "console.log(require.resolve('@prettier/sync'))" --conditions import
node --eval "console.log(require.resolve('@prettier/sync'))" --conditions require
```

Tools that validate **package.json** can also reach it via the new `"./*": "./*"` export too:

```shell
node --eval "console.log(require.resolve('@prettier/sync/package.json'))" --conditions import
node --eval "console.log(require.resolve('@prettier/sync/package.json'))" --conditions require
```

File: index.d.cts (hunk 2 above)
Comment (colinrotherham) [reply index.d.cts]: Let me look into it

It's the `tsc` compiler automatically exporting `@typedef` by default

Comment (fisker) [reply index.d.cts]: How about hand-writing code like this

```ts
import type {
  version,
  util,
  doc,
  formatWithCursor,
  format,
  check,
  resolveConfig,
  resolveConfigFile,
  getFileInfo,
  getSupportInfo,
} from "prettier";

type SynchronizedPrettier = {
  // Prettier static properties
  version: typeof version;
  util: typeof util;
  doc: typeof doc;

  // Prettier functions
  formatWithCursor: Awaited<ReturnType<typeof formatWithCursor>>;
  format: Awaited<ReturnType<typeof format>>;
  check: Awaited<ReturnType<typeof check>>;
  resolveConfig: Awaited<ReturnType<typeof resolveConfig>>;
  resolveConfigFile: Awaited<ReturnType<typeof resolveConfigFile>>;
  getFileInfo: Awaited<ReturnType<typeof getFileInfo>>;
  getSupportInfo: Awaited<ReturnType<typeof getSupportInfo>>;
};
```

File: index.d.cts (hunk 1 above)
Comment (colinrotherham) [reply index.d.cts]: They're not publicly exported, but I've added `/** @type {const} */` to these two:

```mjs
PRETTIER_ASYNC_FUNCTIONS
PRETTIER_STATIC_PROPERTIES
```

When a function is called or property is accessed that doesn't exist (or renamed) it'll validate against:

```mjs
FunctionName in typeof PRETTIER_ASYNC_FUNCTIONS[number]
PropertyName in typeof PRETTIER_STATIC_PROPERTIES[number]
```

Otherwise we'd have to duplicate them in quite a few places:

```ts
FunctionName extends "formatWithCursor" | "format" | "check" | "resolveConfig" | "resolveConfigFile" | "clearConfigCache" | "getFileInfo" | "getSupportInfo"
Property extends "version" | "util" | "doc"
```

These consts are used to populate the functions/properties available in editors:

Prettier available functions and properties

File: index.d.cts (hunk 2 above)
Comment (colinrotherham) [reply index.d.cts]: Entirely up to you 😊

With JSDoc and `checkJs` enabled (without switching to TypeScript) you'll be able to use code completion, see documentation in tooltips, and spot and errors and warnings as-you-type—in this repo too

Plus it lets the compiler validate that type declarations are correct when [**`prettier`** `"./src/index.d.ts"` types](https://github.com/prettier/prettier/blob/main/package.json#L17C17-L17C33) change without having to keep types synchronised between repos by hand

### tsconfig.json
```json
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true
  }
}
```

I'd be more than happy if you used this PR for guidance to add them manually 👍

Let me know if you'd like anything splitting out

**Update:** Split out manually in https://github.com/prettier/prettier-synchronized/pull/10/commits/7d922c2e631180f7e66f093da1f94a842fda249a

File: index.d.cts (hunk 1 above)
Comment (colinrotherham) [reply index.d.cts]: Appreciate they're not needed externally so tried adding `@internal`

But when setting **tsconfig.json** `"stripInternal": true` it didn't work due to:

* https://github.com/microsoft/TypeScript/issues/46407
* https://github.com/microsoft/TypeScript/issues/46515

One of the downsides to using the compiler with JSDoc 😔

**Comment by fisker** [comment]
I have no experience with type declarations.

@SimenB Can you help review?

File: package.json (hunk 3 above)
Comment (remcohaszing) [reply package.json]: The type definitions are co-located with the JavaScript file. (`index.cjs` / `index.d.cts`). This means the `types` field isn’t needed. The entire `exports` can be reverted.

Also additional unrelated exports have been added.

Comment (colinrotherham) [reply package.json]: Sounds good, I wasn't sure if every type aware bundler or runtime would still find the types 👍

Removing `"exports"` entirely prevents `./package.json` + `./index.cjs` being resolved directly, which can cause issues with linters that validate package exports, or TypeScript [`{ "resolvePackageJsonExports": true }`](https://www.typescriptlang.org/tsconfig#resolvePackageJsonExports)

I've split out separate commits for package resolution issues

Happy to move them to separate PR if hand-written types are preferred

**Comment by colinrotherham** [comment]
@fisker I've removed all the unnecessary autogenerated lines from `index.d.cts`

We're left with the good stuff like `Awaited<ReturnType<Prettier[FunctionName]>>`

But I've kept the **tsconfig.json** and JSDoc comments so you get type checking locally too

**Review by fisker — APPROVED** [review]

**Comment by fisker** [comment]
Thank you @colinrotherham 

Released https://github.com/prettier/prettier-synchronized/releases/tag/v0.3.0

**Comment by colinrotherham** [comment]
Brilliant, thanks @fisker

Just a reminder that manually adding the types [without a type declaration map](https://github.com/prettier/prettier-synchronized/pull/10#issuecomment-1630253090) means editor and IDE "Go to definition" will jump to the types rather than the original Prettier source code, should you want that in future

Glad I could help though, appreciate it
